### DISCUSSION TITLE
Optional indentions and optional nested lists

### DISCUSSION THREAD

**Discussion by Piliuta** [facebook/lexical#2680] [Q&A]
Two related questions:
### Is there a way to make indentations optional?
I would like to disable indentations in my app altogether. I want that pressing `Tab` to do nothing (or has default behavior - focus next tab element on the page)
Is there an option to disable indentation?

### Is there a way to limit or disable nested lists?
I want to be able to limit how many levels lists can have (within the editor instance, not for each list independently). Limit to 3 or 5 levels, or even disable to just 1 level. This would probably mean disabling indentation, but also I do not want pasted nested lists to become nested lists either, make them "whatever limit is set" levels, so something in the list node needs to be set up?

**Comment by GermanJablo** [comment]
If you need to use richt text and don't want indentation at all, you should register an OUTDENT_CONTENT_COMMAND and INDENT_CONTENT_COMMAND commands that do nothing and have a higher precedence than those commands. (I would avoid overwriting the tab command only since it can also be indented with the toolbar button for example).

If you want to limit the maximum indentation in lists, take a look at the ListMaxIndentPlugin plugin.

**Comment by Piliuta** [comment]
@EgonBolton thanks so much. Registering these commands does the trick. I've got a code like this:

```javascript
editor.registerCommand(
  INDENT_CONTENT_COMMAND,
  () => true,
  COMMAND_PRIORICOTY_CRITICAL,
),
editor.registerCommand(
  OUTDENT_CONTENT_COMMAND,
  () => true,
  COMMAND_PRIORICOTY_CRITICAL,
),
```

_I couldn't find documentation of what expected usage of the priorities or recommendations, but I guess it can be lower than `COMMAND_PRIORITY_CRITICAL` if I do not have other registers._

> If you want to limit the maximum indentation in lists, take a look at the ListMaxIndentPlugin plugin.

Could you clarify what you mean by the `ListMaxIndentPlugin` plugin? I can not find it anywhere. Did you mean a plugin with other name or you are proposing that I create it?

**Reply by GermanJablo** [reply]
> I couldn't find documentation of what expected usage of the priorities or recommendations, but I guess it can be lower than COMMAND_PRIORITY_CRITICAL if I do not have other registers.

https://github.com/facebook/lexical/blob/main/packages/lexical/src/LexicalEditor.ts#L167

> Could you clarify what you mean by the ListMaxIndentPlugin plugin? I can not find it anywhere. Did you mean a plugin with other name or you are proposing that I create it?

https://github.com/facebook/lexical/blob/main/packages/lexical-playground/src/plugins/ListMaxIndentLevelPlugin/index.ts

Sorry, it was ListMaxIndentLevelPlugin.
I recommend that you clone Lexical and use the search engine. The code is mostly very easy to understand. Good luck!

**Reply by Piliuta** [reply]
> > I couldn't find documentation of what expected usage of the priorities or recommendations, but I guess it can be lower than COMMAND_PRIORITY_CRITICAL if I do not have other registers.
> 
> https://github.com/facebook/lexical/blob/main/packages/lexical/src/LexicalEditor.ts#L167
> 
> > Could you clarify what you mean by the ListMaxIndentPlugin plugin? I can not find it anywhere. Did you mean a plugin with other name or you are proposing that I create it?
> 
> https://github.com/facebook/lexical/blob/main/packages/lexical-playground/src/plugins/ListMaxIndentLevelPlugin/index.ts
> 
> Sorry, it was ListMaxIndentLevelPlugin. I recommend that you clone Lexical and use the search engine. The code is mostly very easy to understand. Good luck!

Ah I see what you mean, the idea is based on the same event register.
https://github.com/facebook/lexical/blob/main/packages/lexical-playground/src/plugins/ListMaxIndentLevelPlugin/index.ts#L78-L84

It still doesn't limit indentions when lists are pasted from other editors (or even from itself)
I am able to create more levels than allowed (7) levels such way.

<img width="739" alt="image" src="https://user-images.githubusercontent.com/1575198/180465526-01c81e57-5ee6-49d5-9c09-2a3bcff82b14.png">

I think in this case I should handle "paste" events too and check lists data in there. Which makes sense and falls back to the same answer - use command listeners.
